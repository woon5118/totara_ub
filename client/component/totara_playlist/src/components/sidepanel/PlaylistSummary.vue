<!--
  This file is part of Totara Enterprise Extensions.

  Copyright (C) 2020 onwards Totara Learning Solutions LTD

  Totara Enterprise Extensions is provided only to Totara
  Learning Solutions LTD's customers and partners, pursuant to
  the terms and conditions of a separate agreement with Totara
  Learning Solutions LTD or its affiliate.

  If you do not have an agreement with Totara Learning Solutions
  LTD, you may not access, use, modify, or distribute this software.
  Please contact [licensing@totaralearning.com] for more information.

  @author Kian Nguyen <kian.nguyen@totaralearning.com>
  @author Alvin Smith <alvin.smith@totaralearning.com>
  @module totara_playlist
-->

<template>
  <div class="tui-playlistSummary">
    <InlineEditing
      v-if="!editing"
      :full-width="true"
      :update-able="updateAble"
      :button-aria-label="$str('adddescription', 'totara_playlist')"
      @click="editing = true"
    >
      <template v-slot:content>
        <div
          v-if="summary"
          ref="content"
          class="tui-playlistSummary__content"
          v-html="summary"
        />
        <div
          v-else-if="!summary && updateAble"
          class="tui-playlistSummary__placeholder"
        >
          {{ $str('adddescription', 'totara_playlist') }}
        </div>
      </template>
    </InlineEditing>

    <Form v-else>
      <UnsavedChangesWarning
        v-if="!content.isEmpty && !submitting"
        :value="content"
      />
      <Weka
        v-model="content"
        :usage-identifier="{
          component: 'totara_playlist',
          area: 'summary',
          instanceId: instanceId,
        }"
        :compact="true"
        variant="totara_playlist-summary"
        :placeholder="$str('adddescription', 'totara_playlist')"
        class="tui-playlistSummary__editor"
      />
      <DoneCancelGroup
        :loading="submitting"
        :disabled="submitting"
        @done="submit"
        @cancel="editing = false"
      />
    </Form>
  </div>
</template>

<script>
import tui from 'tui/tui';
import { FORMAT_JSON_EDITOR } from 'tui/format';
import Weka from 'editor_weka/components/Weka';
import WekaValue from 'editor_weka/WekaValue';

import DoneCancelGroup from 'totara_engage/components/buttons/DoneCancelGroup';
import InlineEditing from 'totara_engage/components/form/InlineEditing';
import UnsavedChangesWarning from 'totara_engage/components/form/UnsavedChangesWarning';
import getPlaylist from 'totara_playlist/graphql/get_playlist_raw';
import updatePlaylist from 'totara_playlist/graphql/update_playlist_summary';

export default {
  components: {
    DoneCancelGroup,
    InlineEditing,
    Weka,
    UnsavedChangesWarning,
  },

  props: {
    instanceId: {
      type: [String, Number],
      required: true,
    },

    summary: {
      type: String,
      default: null,
    },

    updateAble: {
      type: Boolean,
      default: false,
    },
  },

  data() {
    return {
      editing: false,
      submitting: false,
      content: WekaValue.empty(),
    };
  },

  apollo: {
    playlist: {
      query: getPlaylist,
      fetchPolicy: 'network-only',
      skip() {
        return !this.editing;
      },
      variables() {
        return {
          id: this.instanceId,
        };
      },

      result({ data: { playlist } }) {
        if (playlist) {
          if (playlist.summary) {
            this.content = WekaValue.fromDoc(JSON.parse(playlist.summary));
          }
        }
      },
    },
  },
  mounted() {
    this.$_scan();
  },

  updated() {
    this.$_scan();
  },

  methods: {
    handleUpdate(opt) {
      this.$_readJson(opt);
    },

    submit() {
      this.submitting = true;

      this.$apollo
        .mutate({
          mutation: updatePlaylist,
          variables: {
            id: this.instanceId,
            summary: JSON.stringify(this.content.getDoc()),
            summary_format: FORMAT_JSON_EDITOR,
          },

          /**
           *
           * @param {DataProxy} proxy
           * @param {Object} data
           */
          updateQuery: (proxy, data) => {
            proxy.writeQuery({
              query: getPlaylist,
              variables: {
                id: this.instanceId,
              },
              data,
            });
          },
        })
        .finally(() => {
          this.editing = false;
          this.submitting = false;
        });
    },

    $_scan() {
      if (!this.$refs.content) {
        return;
      }

      let element = this.$refs.content;
      tui.scan(element);
    },
  },
};
</script>

<lang-strings>
  {
    "totara_playlist": [
      "adddescription"
    ]
  }
</lang-strings>

<style lang="scss">
.tui-playlistSummary {
  margin: var(--gap-8) 0;

  > p {
    margin: 0;
  }

  &__placeholder {
    color: var(--color-state);
  }
}
</style>
