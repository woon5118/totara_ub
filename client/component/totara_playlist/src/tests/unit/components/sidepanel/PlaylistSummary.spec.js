/**
 * This file is part of Totara Enterprise Extensions.
 *
 * Copyright (C) 2020 onwards Totara Learning Solutions LTD
 *
 * Totara Enterprise Extensions is provided only to Totara
 * Learning Solutions LTD's customers and partners, pursuant to
 * the terms and conditions of a separate agreement with Totara
 * Learning Solutions LTD or its affiliate.
 *
 * If you do not have an agreement with Totara Learning Solutions
 * LTD, you may not access, use, modify, or distribute this software.
 * Please contact [licensing@totaralearning.com] for more information.
 *
 * @author Kian Nguyen <kian.nguyen@totaralearning.com>
 * @module totara_playlist
 */

import Vue from 'vue';
import PlaylistSummary from 'totara_playlist/components/sidepanel/PlaylistSummary';
import { mount } from '@vue/test-utils';

Vue.directive('focus-within', {});

jest.mock('tui/tui', function() {
  return {
    scan() {
      return 'content';
    },
  };
});

describe('totara_playlist/components/sidepanel/PlaylistSummary', () => {
  let wrapper;

  beforeAll(() => {
    wrapper = mount(PlaylistSummary, {
      propsData: {
        instanceId: 1,
        summary: 'Hello world',
      },
    });
  });

  it('Checks snapshot', () => {
    expect(wrapper.element).toMatchSnapshot();
  });
});
