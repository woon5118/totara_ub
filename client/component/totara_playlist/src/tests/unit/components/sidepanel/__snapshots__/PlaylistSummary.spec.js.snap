// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`totara_playlist/components/sidepanel/PlaylistSummary Checks snapshot 1`] = `
<div
  class="tui-playlistSummary"
>
  <div
    class="tui-inlineEditing tui-inlineEditing--fullWidth"
  >
    <div
      class="tui-playlistSummary__content"
    >
      Hello world
    </div>
     
    <button
      aria-label="[[adddescription, totara_playlist]]"
      class="tui-iconBtn tui-inlineEditing__btn tui-iconBtn--small tui-iconBtn--transparent"
      style="display: none;"
      title="[[adddescription, totara_playlist]]"
      type="button"
    >
      <span
        class="tui-iconBtn__wrap"
      >
        <span
          class="tui-iconBtn__label"
        >
          <span
            aria-hidden="true"
            class="tui-iconBtn__icon"
          >
            <span
              class=""
            />
          </span>
           
          <!---->
        </span>
         
        <!---->
         
        <!---->
      </span>
    </button>
  </div>
</div>
`;
